stl/inc/xstring

```diff
@@ -82,17 +105,60 @@ struct _Char_traits { // properties of a string or stream element
         return nullptr;
     }
 
-    static _Elem* move(_Out_writes_all_(_Count) _Elem* const _First1, _In_reads_(_Count) const _Elem* _First2,
-        size_t _Count) noexcept /* strengthened */ {
+    static _CONSTEXPR20 _Elem* move(_Out_writes_all_(_Count) _Elem* const _First1,
+        _In_reads_(_Count) const _Elem* const _First2, const size_t _Count) noexcept /* strengthened */ {
         // copy [_First2, _First2 + _Count) to [_First1, ...), allowing overlap
-        return static_cast<_Elem*>(_CSTD memmove(_First1, _First2, _Count * sizeof(_Elem)));
+#if _HAS_MEMCPY_MEMMOVE_INTRINSICS
+        __builtin_memmove(_First1, _First2, _Count * sizeof(_Elem));
+#else // ^^^ _HAS_MEMCPY_MEMMOVE_INTRINSICS ^^^ / vvv !_HAS_MEMCPY_MEMMOVE_INTRINSICS vvv
+#ifdef __cpp_lib_is_constant_evaluated
+        if (_STD is_constant_evaluated()) {
+            // dest: [_First1, _First1 + _Count)
+            // src: [_First2, _First2 + _Count)
+            // We need to handle overlapping ranges.
+            // If _First1 is in the src range, we need a backward loop.
+            // Otherwise, the forward loop works (even if the back of dest overlaps the front of src).
+
+            if (_First1 == _First2) {
+                return _First1; // Self-assignment; either loop would work, but returning immediately is faster.
+            }
+
+            // Usually, we would compare pointers with less-than, even though they could belong to different arrays.
+            // However, we're not allowed to do that during constant evaluation, so we need a linear scan for equality.
+            bool _Loop_forward = true;
+
+            for (const _Elem* _Src = _First2; _Src != _First2 + _Count; ++_Src) {
+                if (_First1 == _Src) {
+                    _Loop_forward = false;
+                    break;
+                }
+            }
+
+            if (_Loop_forward) {
+                for (size_t _Idx = 0; _Idx < _Count; ++_Idx) {
+                    _First1[_Idx] = _First2[_Idx];
+                }
+            } else {
+                for (size_t _Idx = 0; _Idx < _Count; ++_Idx) {
+                    _First1[_Count - 1 - _Idx] = _First2[_Count - 1 - _Idx];
+                }
+            }
+
+            return _First1;
+        }
+#endif // __cpp_lib_is_constant_evaluated
+
+        _CSTD memmove(_First1, _First2, _Count * sizeof(_Elem));
+#endif // ^^^ !_HAS_MEMCPY_MEMMOVE_INTRINSICS ^^^
+
+        return _First1;
     }
 
-    static _Elem* assign(_Out_writes_all_(_Count) _Elem* const _First, size_t _Count, const _Elem _Ch) noexcept
+    static _CONSTEXPR20 _Elem* assign(
+        _Out_writes_all_(_Count) _Elem* const _First, size_t _Count, const _Elem _Ch) noexcept
     /* strengthened */ {
         // assign _Count * _Ch to [_First, ...)
-        _Elem* _Next = _First;
-        for (; 0 < _Count; --_Count, ++_Next) {
+        for (_Elem* _Next = _First; _Count > 0; --_Count, ++_Next) {
             *_Next = _Ch;
         }

```

Comment:
Since this pattern is pre-existing in for this function, I'm sure there is a reason, but it isn't obvious to me.  Why does this function not follow the previous functions' patterns of calling the `_CSTD mem*` functions if not evaluated in `constexpr` context?  Is it a subtlety of the definitions or edge cases of the functions?

Reply:
Good catch! I stopped here and started typing this comment. The reasoning is that `memset` only works for byte-sized characters, and this `_Char_traits` class template supports char-like types with size greater than one. `_Narrow_char_traits` is used for char-like types of size 1 - notably `char` and `char8_t` - and its `assign` does call `memset` on line 511.

Reply:
The subtlety is that `assign()` for arbitrary `_Elem` has to repeatedly copy that element - which could be 4-byte `char32_t`, or (after centralizing the code as you identified in your other comment) 2-byte `char16_t`, or 1-byte `char`, etc. There's no 4-byte counterpart to `wmemset()` and `memset()`. `copy()` and `move()` differ because the width of the character doesn't matter when copying/moving byte sequences around, we just need `_Count * sizeof(_Elem)` to handle that variation.

Do you think this is worth commenting?

Reply:
To me it is worth commenting, since the other functions follow such a clear pattern and this one notably differs.  If you're making the other change anyway, it seems worth it to me to also throw a comment in there.

Reply:
After thinking about this, I think the things that make this more mysterious are (1) the verbose workarounds needed until we get `__builtin_memcpy`/`__builtin_memmove` support in all compilers, and (2) the fact that `copy` and `move` appear with `find` in the middle (!) and `assign` afterwards. If the intrinsic/workaround `copy/move` appeared together, then all of the "plain loop" functions, the change in patterns would be less weird. So I think this calls for both a reordering *and* a comment.

Reply:
Ok, I've added:

```
    // For copy/move, we can uniformly call memcpy/memmove (or their builtin versions) for all element types.
```

followed by:

```
    // For compare/length/find/assign, we can't uniformly call CRT functions (or their builtin versions).
    // 8-bit: memcmp/strlen/memchr/memset; 16-bit: wmemcmp/wcslen/wmemchr/wmemset; 32-bit: nonexistent
```